libia2/include/ia2.h

```diff
@@ -264,13 +254,86 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+  /* Allocate and protect the stack for this thread's \p i'th compartment. */  \
+  /* Stores the stack in the appropriate slot of ia2_stackptrs. */             \
+                                                                               \
+  /* I'm worried about interference from other threads if any sensitive */     \
+  /* locals of in this function get spilled to the stack. We probably need */  \
+  /* to write this in assembly. */                                             \
+  static void allocate_compartment_stack(size_t i) {                           \
+    extern __thread size_t ia2_compacted_tid;                                  \
+    /* We must validate compartment and thread id to avoid providing */        \
+    /* a write-what-where primitive. */                                        \
+    if (i > 15) {                                                              \
+      printf("invalid compartment id\n");                                      \
+      exit(1);                                                                 \
+    }                                                                          \
+    /* Other compartments could (sadly) modify tid, so only validate/use a */  \
+    /* private copy to avoid TOCTTOU. */                                       \
+    size_t tid = ia2_compacted_tid;                                            \
+    if (tid >= IA2_MAX_THREADS) {                                              \
+      printf("invalid thread id\n");                                           \
+      exit(1);                                                                 \
+    }                                                                          \
+    size_t stack_idx = (i * PAGE_SIZE / sizeof(void *)) + tid;                 \
+    char *stack = allocate_stack(i);                                           \
+    protect_stack(i, stack);                                                   \
+                                                                               \
+    /* Each stack frame start + 8 is initially 16-byte aligned. */             \
+    char *stack_ptr_value = stack + STACK_SIZE - 8;                            \
+                                                                               \
+    /* Save the current pkru. */                                               \
+    uint32_t old_pkru = 0;                                                     \
+    __asm__ volatile(IA2_RDPKRU : "=a"(old_pkru) : "a"(0), "d"(0), "c"(0));    \
+    /* Switch to the compartment's pkru to access its stack pointer. */        \
+    uint32_t new_pkru = ~((3 << (2 * i)) | 3);                                 \
+    __asm__ volatile(IA2_WRPKRU : : "a"(new_pkru), "d"(0), "c"(0));            \
+    /* Forbid overwriting an existing stack. */                                \
+    if (ia2_stackptrs[stack_idx] != NULL) {                                    \
```

Comment:
Don't we need to explicitly initialize `ia2_stackptrs`?

Reply:
We don't, because it has static storage duration, but we could for clarity: https://stackoverflow.com/questions/16015656/are-global-variables-always-initialized-to-zero-in-c